pkg/metrics/metrics.go

```diff
@@ -0,0 +1,53 @@
+package metrics
+
+import (
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	log = logf.Log.WithName("metrics").WithName("metrics")
+)
+
+const (
+	NSXOperatorSubsystem = "nsxoperator"
+	HealthKey            = "nsxoperator_health"
+	ScrapeTimeout        = 30
+)
+
+var (
+	NSXOperatorHealthStats = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: NSXOperatorSubsystem,
+			Name:      HealthKey,
+			Help:      "Last health status for NSX-Operator. 1 for 'status' label with current status.",
+		},
+		[]string{"status"},
+	)
+)
+
+var registerMetrics sync.Once
+
+// Register all metrics.
+func register(registry *prometheus.Registry) {
+	registerMetrics.Do(func() {
+		registry.MustRegister(NSXOperatorHealthStats)
+	})
+}
+
+// RegisterMetricsAndInstallHandler registers the NSX Operator metrics
+// objects with Prometheus and installs the Prometheus http handler at the
+// default context.
+func RegisterMetricsAndInstallHandler(m *http.ServeMux, metricsAddr string) {
+	registry := prometheus.NewRegistry()
+	register(registry)
+	m.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}))
+	address := net.JoinHostPort("", metricsAddr)
+	http.ListenAndServe(address, m)
+	log.Info("Registered /metrics with prometheus")
+}
```

Comment:
kubebuilder has exposed the metrics port 8080 by default so I think we can reuse it. You can refer to https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics

Reply:
thanks, Qian. So we just use the metrics server started by controller.